blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ElasticsearchRestHelper.java

```diff
@@ -64,9 +66,11 @@ private ElasticsearchRestHelper(){
         this.closeableHttpClient = HttpClients.custom().setConnectionManager(pool).build();
     }
 
-    private String getBaseUrl(){
-        if(baseUrlIndex >= baseUrlArray.length) baseUrlIndex = 0;
-        return String.format("http://%s", baseUrlArray[baseUrlIndex++]);
+    private String getNextBaseUrl(){
+        if(baseUrlIndex.get() >= baseUrlArray.length) baseUrlIndex.set(0);
+        String nextBaseUrl = String.format("http://%s", baseUrlArray[baseUrlIndex.get()]);
+        baseUrlIndex.set(baseUrlIndex.get()+1);
```

Comment:
This usage defeats the benefit of AtomInteger. Instead use [incrementAndGet](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html#incrementAndGet--) or `getAndIncrement`.

**HOWEVER**, I later realized even using AtomInteger doesn't prevent a race condition where a thread could see and use an index beyond the array length. Sorry for the extra work I caused, but I would recommend switching back to the code you had before **but** declare the method synchronized.

Reply:
This is exactly I saw and was debating on that as you mentioned here. :-)